src/main/java/emissary/place/ServiceProviderRefreshablePlace.java

```diff
@@ -0,0 +1,169 @@
+package emissary.place;
+
+import emissary.config.ConfigUtil;
+import emissary.config.Configurator;
+import emissary.core.Factory;
+
+import jakarta.annotation.Nullable;
+import org.apache.commons.collections4.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * ServiceProviderPlace that supports on-demand refresh of its configuration
+ */
+public abstract class ServiceProviderRefreshablePlace extends ServiceProviderPlace {
+
+    private static final Logger logger = LoggerFactory.getLogger(ServiceProviderRefreshablePlace.class);
+
+    private final Object allocatorLock = new Object();
+    private final AtomicBoolean invalidated = new AtomicBoolean(false);
+    private final AtomicBoolean defunct = new AtomicBoolean(false);
```

Comment:
I'm not sure I understand the intended role of `defunct`, but it seems to limit the place to at-most one refresh, even if `invalidate` is called again.  Is that intentional?

Reply:
Yes this is intentional. You shouldn't be able to get in the state where you can pull a defunct place, but I was trying to future proof the code. When a place is refreshed, a new object is created and swaps out with the old one in the namespace. I didn't want the off chance that we refresh a place more than once.